// appium-logger/build/test/helpers.js
'use strict';

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _chai = require('chai');

var _chai2 = _interopRequireDefault(_chai);

var _sinon = require('sinon');

var _sinon2 = _interopRequireDefault(_sinon);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _libLogger = require('../lib/logger');

_chai2['default'].should();

function setupWriters() {
  return { 'stdout': _sinon2['default'].spy(process.stdout, 'write'),
    'stderr': _sinon2['default'].spy(process.stderr, 'write') };
}

function getDynamicLogger(testingMode, forceLogs) {
  var prefix = arguments.length <= 2 || arguments[2] === undefined ? null : arguments[2];

  process.env._TESTING = testingMode ? '1' : '0';
  process.env._FORCE_LOGS = forceLogs ? '1' : '0';
  return (0, _libLogger.getLogger)(prefix);
}

function restoreWriters(writers) {
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = _getIterator(_lodash2['default'].values(writers)), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var w = _step.value;

      w.restore();
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator['return']) {
        _iterator['return']();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }
}

function someoneHadOutput(writers, output) {
  var hadOutput = false;
  var matchOutput = _sinon2['default'].match(function (value) {
    return value && value.indexOf(output) >= 0;
  }, 'matchOutput');

  var _iteratorNormalCompletion2 = true;
  var _didIteratorError2 = false;
  var _iteratorError2 = undefined;

  try {
    for (var _iterator2 = _getIterator(_lodash2['default'].values(writers)), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
      var writer = _step2.value;

      if (writer.calledWith) {
        hadOutput = writer.calledWithMatch(matchOutput);
        if (hadOutput) break;
      }
    }
  } catch (err) {
    _didIteratorError2 = true;
    _iteratorError2 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion2 && _iterator2['return']) {
        _iterator2['return']();
      }
    } finally {
      if (_didIteratorError2) {
        throw _iteratorError2;
      }
    }
  }

  return hadOutput;
}

function assertOutputContains(writers, output) {
  if (!someoneHadOutput(writers, output)) {
    throw new Error('Expected something to have been called with: \'' + output + '\'');
  }
}

function assertOutputDoesntContain(writers, output) {
  if (someoneHadOutput(writers, output)) {
    throw new Error('Expected nothing to have been called with: \'' + output + '\'');
  }
}

exports.setupWriters = setupWriters;
exports.restoreWriters = restoreWriters;
exports.assertOutputContains = assertOutputContains;
exports.assertOutputDoesntContain = assertOutputDoesntContain;
exports.getDynamicLogger = getDynamicLogger;
