// appium-support/build/lib/plist.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _plist = require('plist');

var _plist2 = _interopRequireDefault(_plist);

var _bplistCreator = require('bplist-creator');

var _bplistCreator2 = _interopRequireDefault(_bplistCreator);

var _bplistParser = require('bplist-parser');

var _bplistParser2 = _interopRequireDefault(_bplistParser);

var _fs = require('./fs');

var _fs2 = _interopRequireDefault(_fs);

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var parseFile = _bluebird2['default'].promisify(_bplistParser2['default'].parseFile);

// XML Plist library helper
function parseXmlPlistFile(plistFilename) {
  var xmlContent;
  return _regeneratorRuntime.async(function parseXmlPlistFile$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(_fs2['default'].readFile(plistFilename, 'utf8'));

      case 2:
        xmlContent = context$1$0.sent;
        return context$1$0.abrupt('return', _plist2['default'].parse(xmlContent));

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function parsePlistFile(plist) {
  var obj;
  return _regeneratorRuntime.async(function parsePlistFile$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        _logger2['default'].debug('Attempting to parse plist file \'' + plist + '\'');
        obj = undefined;
        context$1$0.prev = 2;
        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(parseFile(plist));

      case 5:
        obj = context$1$0.sent;
        context$1$0.next = 23;
        break;

      case 8:
        context$1$0.prev = 8;
        context$1$0.t0 = context$1$0['catch'](2);

        _logger2['default'].debug('Could not parse plist file as binary');
        _logger2['default'].debug('Will try to parse the plist file as XML');
        context$1$0.prev = 12;
        context$1$0.next = 15;
        return _regeneratorRuntime.awrap(parseXmlPlistFile(plist));

      case 15:
        obj = context$1$0.sent;
        context$1$0.next = 21;
        break;

      case 18:
        context$1$0.prev = 18;
        context$1$0.t1 = context$1$0['catch'](12);

        _logger2['default'].errorAndThrow('Could not parse plist file as XML: ' + context$1$0.t1.message);

      case 21:
        _logger2['default'].debug('Parsed file as XML');
        return context$1$0.abrupt('return', obj);

      case 23:
        if (!obj.length) {
          context$1$0.next = 28;
          break;
        }

        _logger2['default'].debug('Parsed plist file as binary');
        return context$1$0.abrupt('return', obj[0]);

      case 28:
        throw new Error('Binary file ' + plist + ' appears to be empty');

      case 29:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[2, 8], [12, 18]]);
}

function updatePlistFile(plist, updatedFields) {
  var binary = arguments.length <= 2 || arguments[2] === undefined ? true : arguments[2];
  var obj, newPlist;
  return _regeneratorRuntime.async(function updatePlistFile$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        _logger2['default'].debug('Attempting to update plist file\'' + plist + '\'');
        obj = undefined;
        context$1$0.prev = 2;
        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(parsePlistFile(plist));

      case 5:
        obj = context$1$0.sent;
        context$1$0.next = 11;
        break;

      case 8:
        context$1$0.prev = 8;
        context$1$0.t0 = context$1$0['catch'](2);

        _logger2['default'].errorAndThrow('Could not update plist: ' + context$1$0.t0.message);

      case 11:
        _lodash2['default'].extend(obj, updatedFields);
        newPlist = binary ? (0, _bplistCreator2['default'])(obj) : _plist2['default'].build(obj);
        context$1$0.prev = 13;
        context$1$0.next = 16;
        return _regeneratorRuntime.awrap(_fs2['default'].writeFile(plist, newPlist));

      case 16:
        context$1$0.next = 21;
        break;

      case 18:
        context$1$0.prev = 18;
        context$1$0.t1 = context$1$0['catch'](13);

        _logger2['default'].errorAndThrow('Could not save plist: ' + context$1$0.t1.message);

      case 21:
        _logger2['default'].debug('Wrote plist file \'' + plist + '\'');

      case 22:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[2, 8], [13, 18]]);
}

exports.parsePlistFile = parsePlistFile;
exports.updatePlistFile = updatePlistFile;
